### ISSUE TITLE
Documentation on Bitbucket Server Setup 

### ISSUE THREAD

**Issue by port22** [bleenco/abstruse#394]
Hi guys,
I'm still having a hard time figuring out how to connect a self-hosted Bitbucket Server with abstruse.

There are some closed issues here but they didn't answer all my questions, so maybe we could work together on a little documentation on how to setup Bitbucket Server.

Things I've done:

- start abstruse docker on the same host as Bitbucket, :6500
- add a personal access token in Bitbucket for my own username (for testing), clipboarded token
- go to abstruse -> config -> access tokens -> bitbucket
  - Description: Bitbucket
  - Bitbucket Client ID: my bitbucket username
  - Bitbucket OAuth Key: unknown?
  - Bitbucket OAuth Secret: unknown?

I assume, abstruse is not meant to be application-linked with Bitbucket, but rather be configured via Web-Hook with an url like http://abstruse-server:6500/webhooks/bitbucket and with the clipboarded token (as I guessed). And this returns 400 per default (because Bitbucket is not set up in abstruse because of the missing fields).

Now, what is OAuth Key&Secret and where do I get it from?

Regarding the Client ID: I want, of course, that any push to the repo by any Bitbucket user will trigger a build. So the Client ID ist just used as the authenticated user under the hood? Not that for every user a token has to be generated ...

**Comment by jkuri** [comment]
* Run `abstruse` behind some proxy and not expose abstruse through port 6500 by default.
* I tried configure bitbucket now myself and this is the following procedure;

1. on Bitbucket goto `user settings` -> `access management` -> `OAuth`
- generate new consumer to look something like that:
![screen shot 2018-08-09 at 23 37 17](https://user-images.githubusercontent.com/1796022/43927766-de98a4d4-9c2e-11e8-8a27-8979c24231f7.png)

2. go to abstruse `user settings` -> `access tokens` -> add Bitbucket credentials
- Bitbucket Client ID: your bitbucket username (not email)
- Bitbucket OAuth Key: check previous screenshot
- Bitbucket OAuth Secret: check previous screenshot

3. on Bitbucket go to repository settings -> webhooks -> Add webhook
- make it look like that:
![screen shot 2018-08-09 at 23 41 41](https://user-images.githubusercontent.com/1796022/43927946-7b356138-9c2f-11e8-9cdc-55b01ee1d40f.png)

4. push new commit to that repository so it will automatically trigger build on abstruse
- stop the build since auth is not set for repository already

5. on Abstruse go to `Repositories` -> `your repository` -> `settings`
- mark access token you set before for this repository:
![screen shot 2018-08-09 at 23 43 49](https://user-images.githubusercontent.com/1796022/43928029-bdc7a024-9c2f-11e8-8850-1b77d2f3651c.png)

6. push another commit on your repository, everything should work now!
![screen shot 2018-08-09 at 23 46 17](https://user-images.githubusercontent.com/1796022/43928048-cd61fa70-9c2f-11e8-9de8-75f816e2af16.png)

I really hope this tutorial is helpful for your since I know that configuration of bitbucket repositories is far from ideal on abstruse.

* To answer your question; So the Client ID is just used as the authenticated user under the hood?
- its the username used to clone repository on abstruse on every build
- only you should generate that oauth token, not every user and it will work for all.

Please let me know if this works for you. Thanks!

**Comment by port22** [comment]
Stuck on point 1):

![screenshot from 2018-08-10 00-21-58](https://user-images.githubusercontent.com/8873017/43929014-7610c6f8-9c33-11e8-92e8-8fb7ad2f2698.png)

That are the options when I click at the very top right on my avatar -> manage account.
On the administration page (/admin) you'll see that:

![screenshot from 2018-08-10 00-24-22](https://user-images.githubusercontent.com/8873017/43929095-c3413e58-9c33-11e8-9974-c9bffd3f16d7.png)

**Comment by jkuri** [comment]
how can I try this myself? do I need a license to check this out?

V pet., 10. avg. 2018 00:24 je oseba port22 <notifications@github.com>
napisala:

**Comment by jkuri** [comment]
I tried installing Bitbucket Server locally and webhooks are just not integrated in the server itself by default, which differs from Bitbucket Cloud. 
I tried installing webhooks plugin but the JSON payload format sended on push/PR differs from the format Bitbucket Cloud is using.

I am sorry to say that but currently the support for Bitbucket Server is not planned. I am not saying that will for sure not happen in the future but I doubt this will be scheduled soon. Thanks for understading.
